cassiopeia/tools/small_parsimony.py

```diff
@@ -0,0 +1,230 @@
+    meta_item: str,
+    root: Optional[str] = None,
+    state_key: str = "S1",
+    label_key: str = "label",
+    copy: bool = False,
+) -> Optional[CassiopeiaTree]:
+    """Run the Fitch-Hartigan algorithm.
+    
+    Performs the full Fitch-Hartigan small parsimony algorithm which, given
+    a set of states for the leaves, infers the most-parsimonious set of states
+    and returns a random solution that satisfies the maximum-parsimony
+    criterion. The solution will be stored in the label key specified by the
+    user (by default 'label'). This function will modify the tree in place
+    if `copy=False`.
+
+    Args:
+        cassiopeia_tree: CassiopeiaTree that has been processed with the
+            Fitch-Hartigan bottom-up algorithm.
+        meta_item: A column in the CassiopeiaTree cell meta corresponding to a
+            categorical variable.
+        root: Root from which to begin this refinement.
+        state_key: Attribute key that stores the Fitch-Hartigan ancestral states.
+        label_key: Key to add that stores the maximum-parsimony assignment
+            inferred from the Fitch-Hartigan top-down refinement.
+        copy: Modify the tree in place or not.
+    
+    Returns:
+        A new CassiopeiaTree if the copy is set to True, else None.
+    """
+
+    cassiopeia_tree = cassiopeia_tree.copy() if copy else cassiopeia_tree
+
+    fitch_hartigan_bottom_up(cassiopeia_tree, meta_item, state_key)
+
+    fitch_hartigan_top_down(
+        cassiopeia_tree, root, state_key, label_key
+    )
+
+    return cassiopeia_tree if copy else None
+
+
+def fitch_hartigan_bottom_up(
+    cassiopeia_tree: CassiopeiaTree,
+    meta_item: str,
+    add_key: str = "S1",
+    copy: bool = False,
+) -> Optional[CassiopeiaTree]:
+    """Performs Fitch-Hartigan bottom-up ancestral reconstruction.
+
+    Performs the bottom-up phase of the Fitch-Hartigan small parsimony
+    algorithm. A new attribute called "S1" will be added to each node
+    storing the optimal set of ancestral states inferred from this bottom-up 
+    algorithm. If copy is False, the tree will be modified in place.
+     
+
+    Args:
+        cassiopeia_tree: CassiopeiaTree object with cell meta data.
+        meta_item: A column in the CassiopeiaTree cell meta corresponding to a
+            categorical variable.
+        add_key: Key to add for bottom-up reconstruction
+        copy: Modify the tree in place or not.
+
+    Returns:
+        A new CassiopeiaTree if the copy is set to True, else None.
+
+    Raises:
+        CassiopeiaError if the tree does not have the specified meta data
+            of the meta data is not categorical.
+    """
+
+    if meta_item not in cassiopeia_tree.cell_meta.columns:
+        raise CassiopeiaError("Meta item does not exist in the cassiopeia tree")
+
+    meta = cassiopeia_tree.cell_meta[meta_item]
+
+    if meta.dtype not in ['str', 'object']:
+        raise CassiopeiaError("Meta item is not a categorical variable.")
```

Comment:
How about `category` and `int` dtypes?
Also, what's the reason for including a generic `object` dtype here?

Reply:
You're right - I should really only be supporting `categories`.

Small-parsimony only works for categorical data, so it should break if you pass it an int or float.